File: lib/tui/border.rb
# frozen_string_literal: true

require_relative "ansi"

module Tui
  # The single owner of box-drawing chrome: rounded corners plus an angled
  # truecolor gradient swept across a container's border cells. Frame, modals,
  # and the palettes all route their boxes through here so every container gets
  # the same look, and so the box-drawing glyphs live in exactly one place.
  #
  # A gradient is { stops: [[r,g,b], …], angle: degrees }. For each border cell
  # the cell is projected onto the angle's direction vector (with a vertical
  # aspect correction, since terminal cells are ~2× taller than wide), the
  # projection is normalized across the box's four corners to t ∈ [0,1], and the
  # stops are interpolated at t. When no gradient is active (no truecolor, a
  # NO_COLOR/mono theme, or none configured) the border falls back to a single
  # solid SGR — same call sites, no branching at the caller.
  module Border
    A = Ansi

    # Rounded outer corners; the divider tees (├ ┤) are shared by both charsets
    # since the header/footer rules keep their square junctions.
    ROUND  = { tl: "╭", tr: "╮", bl: "╰", br: "╯", h: "─", v: "│", ml: "├", mr: "┤" }.freeze
    SQUARE = { tl: "┌", tr: "┐", bl: "└", br: "┘", h: "─", v: "│", ml: "├", mr: "┤" }.freeze

    # Distinct colors along the ring. Quantizing to a fixed palette both caps the
    # number of SGR sequences emitted and lets adjacent equal-color cells coalesce
    # into one run, so a ~200-cell perimeter costs a couple dozen escapes, not 200.
    DEFAULT_STEPS  = 24
    # Terminal cells are about twice as tall as wide; scaling y by this makes a
    # configured angle read on screen as that angle rather than vertically skewed.
    DEFAULT_ASPECT = 2.0

    module_function

    def chars(corners) = corners == :square ? SQUARE : ROUND

    # Truecolor gating. `nil` (the default) auto-detects; tests set it explicitly.
    # Absent truecolor → gradients drop to the solid fallback.
    def truecolor=(value)
      @truecolor = value
    end

    def truecolor?
      return @truecolor unless @truecolor.nil?
      return false if ENV["NO_COLOR"] && !ENV["NO_COLOR"].empty?
      return true if ENV["COLORTERM"].to_s.match?(/truecolor|24bit/i)

      term = ENV["TERM"].to_s
      !(term.empty? || term == "dumb")
    end

    # Parse a border_gradient spec: two-or-more hex stops then "@<angle>", e.g.
    # "#7aa2f7 #bb9af7 @60". Returns { stops:, angle: } or nil if malformed, so a
    # bad config value degrades to the solid border rather than raising.
    def parse_gradient(spec)
      tokens = spec.to_s.strip.downcase.split(/\s+/)
      return nil if tokens.empty?

      angle = 0.0
      stops = []
      tokens.each do |tok|
        if tok.start_with?("@")
          deg = tok[1..]
          return nil unless deg.match?(/\A-?\d+(?:\.\d+)?\z/)

          angle = deg.to_f
        elsif tok.match?(/\A#\h{6}\z/)
          stops << tok[1..].scan(/../).map { |h| h.to_i(16) }
        else
          return nil
        end
      end
      return nil if stops.length < 2

      { stops: stops, angle: angle }
    end

    # Build a Painter for a box of `width` × `height` cells. `gradient` is a
    # parsed spec or nil; `solid` is an SGR opener string (e.g. "\e[90m") or ""
    # used when the gradient is inactive.
    def painter(width:, height:, gradient: nil, solid: "", steps: DEFAULT_STEPS, aspect: DEFAULT_ASPECT)
      Painter.new(width: width, height: height, gradient: gradient, solid: solid, steps: steps, aspect: aspect)
    end

    # Render a complete box: rounded corners, gradient-swept edges, an optional
    # title strip on the top row. `inner_lines` are each exactly `inner_width`
    # visible cells (any inner margin is the caller's); the vertical edges are
    # added here. `title` (already styled by the caller, e.g. :modal_title) sits
    # after `title_lead` leading dashes; the border glyphs around it are painted
    # while the title text passes through untouched. Returns full rows, each
    # `inner_width + 2` cells wide.
    def box(inner_lines:, inner_width:, gradient: nil, solid: "", title: nil, title_lead: 0,
            corners: :round, steps: DEFAULT_STEPS, aspect: DEFAULT_ASPECT)
      width = inner_width + 2
      height = inner_lines.length + 2
      c = chars(corners)
      p = painter(width: width, height: height, gradient: gradient, solid: solid, steps: steps, aspect: aspect)

      rows = [top_row(p, c, inner_width, title, title_lead)]
      inner_lines.each_with_index do |line, i|
        y = i + 1
        rows << p.cell(0, y, c[:v]) + line + p.cell(width - 1, y, c[:v])
      end
      rows << p.run(height - 1, 0, [c[:bl], *Array.new(inner_width, c[:h]), c[:br]])
      rows
    end

    def top_row(painter, c, inner_width, title, title_lead)
      return painter.run(0, 0, [c[:tl], *Array.new(inner_width, c[:h]), c[:tr]]) if title.nil?

      lead = title_lead
      tw = A.vislen(title)
      fill = [inner_width - lead - tw, 0].max
      painter.run(0, 0, [c[:tl], *Array.new(lead, c[:h])]) +
        title +
        painter.run(0, 1 + lead + tw, [*Array.new(fill, c[:h]), c[:tr]])
    end

    # Precomputes the per-cell colorizer for one box size. `cell`/`run` return
    # glyphs wrapped in the appropriate SGR; the caller maps only border glyphs
    # through it, leaving interior content untouched.
    class Painter
      RESET = "\e[0m"

      def initialize(width:, height:, gradient:, solid:, steps:, aspect:)
        @solid = solid.to_s
        gradient = nil unless Border.truecolor?
        if gradient
          setup_gradient(width: width, height: height, gradient: gradient,
                         steps: [steps.to_i, 2].max, aspect: aspect.to_f)
        end
      end

      # SGR opener for the cell at (x, y): a truecolor fg in gradient mode, else
      # the solid fallback (possibly "").
      def sgr_at(x, y)
        return @solid unless @palette

        proj = (x * @cos) + (y * @aspect * @sin)
        t = @range.zero? ? 0.0 : (proj - @pmin) / @range
        @palette[(t * (@palette.length - 1)).round.clamp(0, @palette.length - 1)]
      end

      # A single colored glyph, self-closing.
      def cell(x, y, glyph)
        seq = sgr_at(x, y)
        seq.empty? ? glyph : "#{seq}#{glyph}#{RESET}"
      end

      # A run of consecutive cells on row `y` starting at column `x0`, with equal
      # adjacent colors coalesced to one SGR. Closes with a single reset iff any
      # color was emitted.
      def run(y, x0, glyphs)
        out = +""
        last = nil
        styled = false
        glyphs.each_with_index do |glyph, i|
          seq = sgr_at(x0 + i, y)
          if seq != last
            out << seq
            last = seq
            styled ||= !seq.empty?
          end
          out << glyph
        end
        out << RESET if styled
        out
      end

      private

      def setup_gradient(width:, height:, gradient:, steps:, aspect:)
        rad = gradient[:angle] * Math::PI / 180.0
        @cos = Math.cos(rad)
        @sin = Math.sin(rad)
        @aspect = aspect
        corners = [[0, 0], [width - 1, 0], [0, height - 1], [width - 1, height - 1]]
        projections = corners.map { |cx, cy| (cx * @cos) + (cy * aspect * @sin) }
        @pmin = projections.min
        @range = projections.max - @pmin
        stops = gradient[:stops]
        @palette = (0...steps).map do |i|
          r, g, b = lerp(stops, i / (steps - 1.0))
          "\e[38;2;#{r};#{g};#{b}m"
        end
      end

      # Interpolate the stop list at t ∈ [0,1] in plain sRGB. Predictable and
      # fine for the adjacent-hue sweeps borders use; isolated here so the color
      # space can be upgraded (linear-light / OkLab) without touching callers.
      def lerp(stops, t)
        return stops.first if stops.length == 1

        seg = (t * (stops.length - 1)).clamp(0, stops.length - 1)
        i = seg.floor.clamp(0, stops.length - 2)
        f = seg - i
        a = stops[i]
        b = stops[i + 1]
        [0, 1, 2].map { |k| (a[k] + ((b[k] - a[k]) * f)).round }
      end
    end
  end
end

File: go/internal/tui/term/border/border.go
// Package border is the single owner of box-drawing chrome: rounded corners
// plus an angled truecolor gradient swept across a container's border cells.
// Frames, modals, and palettes all route their boxes through here so every
// container gets the same look, and so the box-drawing glyphs live in exactly
// one place.
//
// A Gradient is a stop list plus an angle in degrees. For each border cell the
// cell is projected onto the angle's direction vector (with a vertical aspect
// correction, since terminal cells are ~2x taller than wide), the projection is
// normalized across the box's four corners to t in [0,1], and the stops are
// interpolated at t. When no gradient is active (no truecolor, a NO_COLOR/mono
// theme, or none configured) the border falls back to a single solid SGR —
// same call sites, no branching at the caller.
//
// Go port of Ruby's lib/tui/border.rb. Truecolor availability is a Painter
// option rather than Ruby's module-level flag.
package border

import (
	"math"
	"regexp"
	"strconv"
	"strings"

	"tasks-go/internal/tui/term/ansi"
)

// Charset is one box-drawing glyph set.
type Charset struct {
	TL, TR, BL, BR, H, V, ML, MR string
}

// Round has rounded outer corners; Square keeps them square. The divider tees
// (ML/MR) are shared by both, since the header/footer rules keep their square
// junctions.
var (
	Round  = Charset{TL: "╭", TR: "╮", BL: "╰", BR: "╯", H: "─", V: "│", ML: "├", MR: "┤"}
	Square = Charset{TL: "┌", TR: "┐", BL: "└", BR: "┘", H: "─", V: "│", ML: "├", MR: "┤"}
)

// Corners selects a charset.
type Corners int

const (
	CornersRound Corners = iota
	CornersSquare
)

// Chars returns the charset for the requested corner style.
func Chars(c Corners) Charset {
	if c == CornersSquare {
		return Square
	}
	return Round
}

const (
	// DefaultSteps is the number of distinct colors quantized along the ring.
	// Quantizing both caps the number of SGR sequences emitted and lets
	// adjacent equal-color cells coalesce into one run, so a ~200-cell
	// perimeter costs a couple dozen escapes, not 200.
	DefaultSteps = 24
	// DefaultAspect scales y so a configured angle reads on screen as that
	// angle rather than vertically skewed.
	DefaultAspect = 2.0
)

const reset = "\x1b[0m"

// RGB is one gradient stop.
type RGB [3]int

// Gradient is a parsed border_gradient spec.
type Gradient struct {
	Stops []RGB
	Angle float64
}

var (
	hexStop  = regexp.MustCompile(`^#[0-9a-f]{6}$`)
	angleNum = regexp.MustCompile(`^-?\d+(?:\.\d+)?$`)
)

// ParseGradient parses a border_gradient spec: two or more hex stops then an
// "@<angle>", e.g. "#7aa2f7 #bb9af7 @60". It returns nil if the spec is
// malformed, so a bad config value degrades to the solid border rather than
// failing.
func ParseGradient(spec string) *Gradient {
	tokens := strings.Fields(strings.ToLower(strings.TrimSpace(spec)))
	if len(tokens) == 0 {
		return nil
	}
	g := &Gradient{}
	for _, tok := range tokens {
		switch {
		case strings.HasPrefix(tok, "@"):
			deg := tok[1:]
			if !angleNum.MatchString(deg) {
				return nil
			}
			f, err := strconv.ParseFloat(deg, 64)
			if err != nil {
				return nil
			}
			g.Angle = f
		case hexStop.MatchString(tok):
			r, _ := strconv.ParseInt(tok[1:3], 16, 32)
			gr, _ := strconv.ParseInt(tok[3:5], 16, 32)
			b, _ := strconv.ParseInt(tok[5:7], 16, 32)
			g.Stops = append(g.Stops, RGB{int(r), int(gr), int(b)})
		default:
			return nil
		}
	}
	if len(g.Stops) < 2 {
		return nil
	}
	return g
}

// Env looks up an environment variable; nil means "no environment".
type Env func(string) string

// DetectTruecolor is the Ruby truecolor probe: a non-empty NO_COLOR disables
// color entirely, COLORTERM naming truecolor/24bit enables it, and otherwise
// any TERM that is neither empty nor "dumb" is assumed capable.
func DetectTruecolor(env Env) bool {
	get := func(k string) string {
		if env == nil {
			return ""
		}
		return env(k)
	}
	if get("NO_COLOR") != "" {
		return false
	}
	colorterm := strings.ToLower(get("COLORTERM"))
	if strings.Contains(colorterm, "truecolor") || strings.Contains(colorterm, "24bit") {
		return true
	}
	term := get("TERM")
	return term != "" && term != "dumb"
}

// PainterOptions configures one box-sized colorizer.
type PainterOptions struct {
	Width, Height int
	Gradient      *Gradient
	// Solid is an SGR opener (e.g. "\e[90m") or "" used when the gradient is
	// inactive.
	Solid string
	// Truecolor gates the gradient. Absent truecolor, gradients drop to Solid.
	Truecolor bool
	Steps     int
	Aspect    float64
}

// Painter precomputes the per-cell colorizer for one box size. Cell and Run
// return glyphs wrapped in the appropriate SGR; the caller maps only border
// glyphs through it, leaving interior content untouched.
type Painter struct {
	solid   string
	palette []string
	cos     float64
	sin     float64
	aspect  float64
	pmin    float64
	rng     float64
}

// NewPainter builds a Painter for a box of Width x Height cells.
func NewPainter(opts PainterOptions) *Painter {
	p := &Painter{solid: opts.Solid}
	if opts.Gradient == nil || !opts.Truecolor {
		return p
	}
	steps := opts.Steps
	if steps == 0 {
		steps = DefaultSteps
	}
	if steps < 2 {
		steps = 2
	}
	aspect := opts.Aspect
	if aspect == 0 {
		aspect = DefaultAspect
	}

	rad := opts.Gradient.Angle * math.Pi / 180.0
	p.cos = math.Cos(rad)
	p.sin = math.Sin(rad)
	p.aspect = aspect
	corners := [4][2]int{{0, 0}, {opts.Width - 1, 0}, {0, opts.Height - 1}, {opts.Width - 1, opts.Height - 1}}
	pmin, pmax := math.Inf(1), math.Inf(-1)
	for _, c := range corners {
		proj := float64(c[0])*p.cos + float64(c[1])*aspect*p.sin
		pmin = math.Min(pmin, proj)
		pmax = math.Max(pmax, proj)
	}
	p.pmin = pmin
	p.rng = pmax - pmin

	p.palette = make([]string, steps)
	for i := 0; i < steps; i++ {
		c := lerp(opts.Gradient.Stops, float64(i)/float64(steps-1))
		p.palette[i] = "\x1b[38;2;" + strconv.Itoa(c[0]) + ";" + strconv.Itoa(c[1]) + ";" + strconv.Itoa(c[2]) + "m"
	}
	return p
}

// SGRAt is the opening sequence for the cell at (x, y): a truecolor foreground
// in gradient mode, else the solid fallback (possibly "").
func (p *Painter) SGRAt(x, y int) string {
	if p.palette == nil {
		return p.solid
	}
	proj := float64(x)*p.cos + float64(y)*p.aspect*p.sin
	t := 0.0
	if p.rng != 0 {
		t = (proj - p.pmin) / p.rng
	}
	index := int(math.Round(t * float64(len(p.palette)-1)))
	if index < 0 {
		index = 0
	}
	if index > len(p.palette)-1 {
		index = len(p.palette) - 1
	}
	return p.palette[index]
}

// Cell renders a single colored glyph, self-closing.
func (p *Painter) Cell(x, y int, glyph string) string {
	seq := p.SGRAt(x, y)
	if seq == "" {
		return glyph
	}
	return seq + glyph + reset
}

// Run renders consecutive cells on row y starting at column x0, coalescing
// equal adjacent colors into one SGR. It closes with a single reset iff any
// color was emitted.
func (p *Painter) Run(y, x0 int, glyphs []string) string {
	var out strings.Builder
	last := ""
	haveLast := false
	styled := false
	for i, glyph := range glyphs {
		seq := p.SGRAt(x0+i, y)
		if !haveLast || seq != last {
			out.WriteString(seq)
			last = seq
			haveLast = true
			if seq != "" {
				styled = true
			}
		}
		out.WriteString(glyph)
	}
	if styled {
		out.WriteString(reset)
	}
	return out.String()
}

// lerp interpolates the stop list at t in [0,1] in plain sRGB. Predictable and
// fine for the adjacent-hue sweeps borders use; isolated here so the color
// space can be upgraded without touching callers.
func lerp(stops []RGB, t float64) RGB {
	if len(stops) == 1 {
		return stops[0]
	}
	seg := t * float64(len(stops)-1)
	if seg < 0 {
		seg = 0
	}
	if seg > float64(len(stops)-1) {
		seg = float64(len(stops) - 1)
	}
	i := int(math.Floor(seg))
	if i > len(stops)-2 {
		i = len(stops) - 2
	}
	if i < 0 {
		i = 0
	}
	f := seg - float64(i)
	a, b := stops[i], stops[i+1]
	var out RGB
	for k := 0; k < 3; k++ {
		out[k] = int(math.Round(float64(a[k]) + (float64(b[k])-float64(a[k]))*f))
	}
	return out
}

// BoxOptions describes one complete box.
type BoxOptions struct {
	// InnerLines are each exactly InnerWidth visible cells (any inner margin is
	// the caller's); the vertical edges are added here.
	InnerLines []string
	InnerWidth int
	Gradient   *Gradient
	Solid      string
	Truecolor  bool
	// Title is already styled by the caller and sits after TitleLead leading
	// dashes; the border glyphs around it are painted while the title text
	// passes through untouched. An empty title draws a plain top rule.
	Title     string
	TitleLead int
	Corners   Corners
	Steps     int
	Aspect    float64
}

// Box renders a complete box. Returned rows are each InnerWidth + 2 cells wide.
func Box(opts BoxOptions) []string {
	width := opts.InnerWidth + 2
	height := len(opts.InnerLines) + 2
	c := Chars(opts.Corners)
	p := NewPainter(PainterOptions{
		Width: width, Height: height, Gradient: opts.Gradient, Solid: opts.Solid,
		Truecolor: opts.Truecolor, Steps: opts.Steps, Aspect: opts.Aspect,
	})

	rows := []string{topRow(p, c, opts.InnerWidth, opts.Title, opts.TitleLead)}
	for i, line := range opts.InnerLines {
		y := i + 1
		rows = append(rows, p.Cell(0, y, c.V)+line+p.Cell(width-1, y, c.V))
	}
	rows = append(rows, p.Run(height-1, 0, glyphRow(c.BL, c.H, opts.InnerWidth, c.BR)))
	return rows
}

func glyphRow(left, fill string, n int, right string) []string {
	out := make([]string, 0, n+2)
	out = append(out, left)
	for i := 0; i < n; i++ {
		out = append(out, fill)
	}
	return append(out, right)
}

func topRow(p *Painter, c Charset, innerWidth int, title string, titleLead int) string {
	if title == "" {
		return p.Run(0, 0, glyphRow(c.TL, c.H, innerWidth, c.TR))
	}
	// Self-enforce the fit: clamp the lead and truncate the (styled) title so
	// lead + title never exceeds innerWidth. Callers already pre-truncate, so
	// this is a no-op for them, but it keeps the top row the same width as the
	// rest of the box for any caller of this shared primitive.
	lead := titleLead
	if lead < 0 {
		lead = 0
	}
	if lead > innerWidth {
		lead = innerWidth
	}
	title = ansi.VTrunc(title, innerWidth-lead)
	tw := ansi.VisLen(title)
	fill := innerWidth - lead - tw

	head := make([]string, 0, lead+1)
	head = append(head, c.TL)
	for i := 0; i < lead; i++ {
		head = append(head, c.H)
	}
	tail := make([]string, 0, fill+1)
	for i := 0; i < fill; i++ {
		tail = append(tail, c.H)
	}
	tail = append(tail, c.TR)
	return p.Run(0, 0, head) + title + p.Run(0, 1+lead+tw, tail)
}
